Add support for NPM credentials.

varungandhi-src opened the pull request:
The largest part of the change is the testing setup. I'm using [Verdaccio](https://github.com/verdaccio/verdaccio) to spin up a proxy for the NPM registry, and testing authorization against that.

Questions and TODOs:
- [X] Should I include any documentation about Verdaccio so that people can change/understand the code better? - Resolution: Removed the usage of Verdaccio altogether.
- [X] Is there a better/simpler way to test this same behavior? - Resolution: Using a mock server which uses a hard-coded JSON response.
- [x] I've left a TODO for adding a test case for an error code path I hit, I will fix that.

efritz commented:
>  Should I include any documentation about Verdaccio so that people can change/understand the code better?

@varungandhi-src I have no clue what `Verdaccio` is so it would probably be appreciated :)

mrnugget commented on internal/extsvc/npm/npm.go:
```diff
@@ -184,8 +186,15 @@ func (client *HTTPClient) makeGetRequest(ctx context.Context, url string) (respo
 	if err != nil {
+		if resp == nil { // possible if you pass in an incorrect registry URL
+			// TODO: [Varun] Add test for this code path.
+			return nil, npmError{-1, err}
```
Do you need to set the `StatusCode` to `-1`? `0` is the default value and also not a valid HTTP status code (iirc).

Feels like it's time to change the method signature to return an `error` and make `npmError` one.

varungandhi-src replied:
Yeah, the -1 isn't needed, I've removed that.

As for changing the error, it seems like the same considerations still apply (this error isn't being passed out of the package), and this change is only adding a new return path to a function which already had 6 return paths returning an `npmError`, so I don't think making `npmError` an `error` is useful.

mrnugget replied:
Except that here `npmError{err}` adds nothing and you could just return `err` if the signature would allow it, right?

varungandhi-src replied:
Added `Error()` method to `npmError` and changed the return types. I was hesitant to make this change but I do admit the code is more uniform after making it.

mrnugget commented on internal/extsvc/npm/npm_test.go:
```diff
@@ -32,13 +37,85 @@ func newTestHTTPClient(t *testing.T) (client *HTTPClient, stop func()) {
+	require.Nil(t, err)
+	defer reader.Close()
+	require.Nil(t, cmd.Start())
+	var allBytes = make([]byte, 4096)
```
Nitpick, but these tests are really hard to visually parse without any whitespace lines in there.

varungandhi-src replied:
Added some newlines to separate out different parts.

mrnugget commented on internal/extsvc/npm/npm_test.go:
```diff
@@ -32,13 +37,85 @@ func newTestHTTPClient(t *testing.T) (client *HTTPClient, stop func()) {
+
+func startLocalNPMRegistry(t *testing.T, hostPort string) (stop func()) {
+	t.Helper()
+	yarnOutput, err := exec.Command("yarn", "bin", "verdaccio").Output()
```
Hmmm. I don't see the huge advantages of using `verdaccio` here (which requires `yarn` to be installed to run `go test`, which is not super nice) vs. writing a simple mock test server with [`httptest.NewServer`](https://pkg.go.dev/net/http/httptest#NewServer) (see [this query](https://sourcegraph.com/search?q=context:global+repo:%5Egithub%5C.com/sourcegraph/%28sourcegraph%7Csrc-cli%29%24+httptest.NewServer+file:batches&patternType=literal) for some examples of test servers I wrote)

As far as I can tell we only test a handful of endpoints that return JSON and set status codes. That would be .. ~50 lines of Go code? instead of having to run external packages with an external package manager.

varungandhi-src replied:
> Hmmm. I don't see the huge advantages of using verdaccio here [..] vs. writing a simple mock test server with httptest.NewServer

The main reason I chose to go this route was that we could test against something that is production-like. The things that is most likely to go wrong here IMO is that some customer is using their own registry with some credentials and it doesn't work for some odd reason. Moreover, the credentials code paths won't be exercised on Sourcegraph.com, because we will be directly using the public NPM registry.

If we test with a mock server, I would have much less confidence that the mock itself was fully correct, compared to a server which is being used in production (Verdaccio seems to be used a fair bit based on the [list of sponsors](https://github.com/verdaccio/verdaccio#open-collective-sponsors)).

> (which requires yarn to be installed to run go test, which is not super nice)

I understand it is a little weird, but:
- We already require Yarn to be installed. https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/dev/sg/sg_setup.go?L322
- It seems like the most portable way to get the path to the `verdaccio` executable.

mrnugget replied:
> The main reason I chose to go this route was that we could test against something that is production-like

How is verdaccio more production-like in regards to the behaviour we're testing here?

As far as I can tell, we do know:
1. what behaviour we want to test
2. which of our code paths we want to exercise
3. what we expect them to do

Doesn't that mean we can't replicate what makes verdaccio more "production-like"?

> The things that is most likely to go wrong here IMO is that some customer is using their own registry with some credentials and it doesn't work for some odd reason.

How would using verdaccio prevent this?

> If we test with a mock server, I would have much less confidence that the mock itself was fully correct, compared to a server which is being used in production (Verdaccio seems to be used a fair bit based on the list of sponsors).

My point was not meant as a general "a mock server is better than using a 3rd party dependency in our tests". I was specifically talking about the tests we have here, which are 2-3 HTTP calls of which we assert their status codes/response bodies.

In that case, I personally would have more confidence in the tests and how reliable they are when they use a mock server that complies with the observed behaviour of the registry vs. a 3rd party dependency that needs to be installed and run and updated

varungandhi-src replied:
I've replaced the code with a mock server returning hard-coded JSON responses.

mrnugget commented on internal/extsvc/npm/npm_test.go:
```diff
@@ -32,13 +37,85 @@ func newTestHTTPClient(t *testing.T) (client *HTTPClient, stop func()) {
+	defer reader.Close()
+	require.Nil(t, cmd.Start())
+	var allBytes = make([]byte, 4096)
+	for i := 0; true; i++ {
```
```suggestion
	for i := 0; ; i++ {
```

mrnugget commented on internal/extsvc/npm/npm.go:
```diff
@@ -100,9 +100,10 @@ type HTTPClient struct {
 }
 
-func NewHTTPClient(registryURL string, rateLimit *schema.NPMRateLimit) *HTTPClient {
+func NewHTTPClient(registryURL string, rateLimit *schema.NPMRateLimit, credentials string) *HTTPClient {
```
Curious: why not use the `auth` package?

varungandhi-src replied:
It seemed simpler to use a `string` directly, since the only use of that `credentials` field is for setting a single header.

varungandhi-src commented on internal/extsvc/npm/npm_test.go:
```diff
@@ -32,13 +37,105 @@ func newTestHTTPClient(t *testing.T) (client *HTTPClient, stop func()) {
+	return string(match)
+}
+
+// [NOTE: verdaccio-usage]
+// We spin up a proxy registry server using Verdaccio (https://verdaccio.org/)
+// for testing credential support in a production-like setting.
+//
+// In case the credentials need to be regenerated in the future for some reason,
+// you can delete the existing config:
+//   $ rm -rf npmtest/{.npmrc,htpasswd,verdaccio-storage}
+// After that, start verdaccio:
+//   $ (cd npmtest; verdaccio --config verdaccio-config.yaml --listen localhost:4981)
+// And create a new user manually with a placeholder name, password and email:
+//   $ NPM_CONFIG_REGISTRY=http://localhost:4981 NPM_CONFIG_USERCONFIG=npmtest/.npmrc npm adduser
+// This should recreate the deleted files, and the storage will fill up as you
+// send requests, such as:
+//   $ curl http://localhost:4981/left-pad/1.3.0 -H "Authorization: Bearer $ACCESS_TOKEN"
+// the ACCESS_TOKEN can be read from the .npmrc file.
```
@efritz I added some docs -- does this look okay to you?

efritz replied:
Thank you!

mrnugget commented on internal/extsvc/npm/npm_test.go:
```diff
@@ -28,20 +31,77 @@ func TestMain(m *testing.M) {
-	client.doer = doer
+	client = NewHTTPClient("https://registry.npmjs.org", &rateLimit, "")
+	stop = func() {}
+	if useRecordReplay {
```
Seems like the test below is the only one where this is `false`. I'd personally pull out the 

```go
client := NewHTTPClient("https://registry.npmjs.org", nil, "")
```

and use that in the test directly. Makes all of this a bit easier to understand.

mrnugget commented on internal/extsvc/npm/npm_test.go:
```diff
@@ -28,20 +31,77 @@ func TestMain(m *testing.M) {
 
+func mockNPMServer(credentials string) *httptest.Server {
+	return httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+		fmt.Printf("req.URL.Path = %s\n", req.URL.Path)
```
Either remove the debug printing, or pass in `*testing.T` to `mockNPMServer` and then use `t.Logf`

```suggestion
```

mrnugget commented on internal/extsvc/npm/npm_test.go:
```diff
@@ -28,20 +31,77 @@ func TestMain(m *testing.M) {
+func mockNPMServer(credentials string) *httptest.Server {
+	return httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+		fmt.Printf("req.URL.Path = %s\n", req.URL.Path)
+		if key, ok := req.Header["Authorization"]; ok && key[0] == fmt.Sprintf("Bearer %s", credentials) {
```
Nitpick: I'd check for `key[0] != ...` and do an early exit here by pulling the `else` branch up. (In request handlers I personally think it's easier to understand the code if it's structured in a sequential-checks way where you can read from top to bottom which conditions a request has to fulfill.)

mrnugget commented on internal/extsvc/npm/npm_test.go:
```diff
@@ -28,20 +31,77 @@ func TestMain(m *testing.M) {
+	return httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+		fmt.Printf("req.URL.Path = %s\n", req.URL.Path)
+		if key, ok := req.Header["Authorization"]; ok && key[0] == fmt.Sprintf("Bearer %s", credentials) {
+			if req.URL.Path == "/left-pad/1.3.1" {
+				w.WriteHeader(http.StatusNotFound)
+				w.Write([]byte(`"version not found: 1.3.1"`))
+			} else if req.URL.Path == "/left-pad/1.3.0" {
+				w.WriteHeader(http.StatusOK)
+				w.Write([]byte(`{"name":"left-pad","dist": {"tarball": "https://registry.npmjs.org/left-pad/-/left-pad-1.3.0.tgz"}}`))
+			} else {
+				panic("Unexpected request!")
+			}
```
Nit/suggestion: you could define some basic routes outside the closure you're returning here

```go
routes := map[string]struct {
	status int
	body   string
}{
	"/left-pad/1.3.1": {
		status: http.StatusNotFound,
		body:   `"version not found: 1.3.1"`,
	},
	"/left-pad/1.3.0": {
		status: http.StatusOK,
		body:   `{"name":"left-pad","dist": {"tarball": "https://registry.npmjs.org/left-pad/-/left-pad-1.3.0.tgz"}}`,
	},
}
```

and then inside the closure you can do

```go
resp, ok := routes[req.URL.Path]
if !ok {
  panic(fmt.Sprintf("unexpected requerst to %s", req.URL.Path))
}
w.WriteHeader(resp.status)
w.Write([]byte(resp.body))
```

I pesonally think that's a bit cleaner and doesn't cost anything expect a bit of typing :)

mrnugget commented on internal/extsvc/npm/npm_test.go:
```diff
@@ -28,20 +31,77 @@ func TestMain(m *testing.M) {
+	client.credentials = "incorrect_credentials"
+
+	_, err = client.DoesDependencyExist(ctx, *presentDep)
+	require.NotNil(t, err)
```
Can you assert here that you get an authorization error? Because the test could pass by simply doing `return errors.New("foo")`, which is not what you want to assert here.

mrnugget commented on internal/extsvc/npm/npm_test.go:
```diff
@@ -103,3 +163,14 @@ func TestFetchSources(t *testing.T) {
 }
+
+func TestNoPanicOnNonexistentRegistry(t *testing.T) {
+	ctx := context.Background()
+	client, stop := newTestHTTPClient(t, true)
+	defer stop()
+	client.registryURL = "http://not-an-npm-registry.sourcegraph.com"
+	dep, err := reposource.ParseNPMDependency("left-pad@1.3.0")
+	require.Nil(t, err)
+	_, err = client.DoesDependencyExist(ctx, *dep)
+	require.NotNil(t, err)
```
Give it some room to breath 😄 

```suggestion
	ctx := context.Background()
	
	client, stop := newTestHTTPClient(t, true)
	defer stop()
	
	client.registryURL = "http://not-an-npm-registry.sourcegraph.com"
	
	dep, err := reposource.ParseNPMDependency("left-pad@1.3.0")
	require.Nil(t, err)
	
	_, err = client.DoesDependencyExist(ctx, *dep)
	require.NotNil(t, err)
```

mrnugget commented on internal/extsvc/npm/npmtest/.gitignore:
```diff
@@ -0,0 +1,2 @@
+verdaccio-storage/
```
Can be removed now, right?

mrnugget commented on internal/extsvc/npm/testdata/vcr/TestCredentials.yaml:
```diff
@@ -0,0 +1,75 @@
+---
```
Do we need this? I would say no, because it doesn't make a lot of sense to replay mock server reponses?

mrnugget commented on package.json:
```diff
@@ -316,6 +316,7 @@
     "typed-scss-modules": "^4.1.1",
     "typescript": "^4.5.2",
     "utc-version": "^2.0.2",
+    "verdaccio": "^5.5.0",
```
Can be removed

varungandhi-src commented:
Cleaned up the code based on your suggestions and squashed changes. Thanks for the thorough review; I should've picked up on some of these things myself before pushing.
